# Data Manipulation with Pandas  
1. Transforming DataFrames
2. Aggregating DataFrames
3. Slicing and Indexing DataFrames
4. Creating and Visualizing DataFrames

### 1. Transforming DataFrames  
  
**What´s the point of pandas?**  
* Data Manipulation
 * Ch 1, Sorting and Subsetting, Creating new columns
 * Ch 2, Summary statistics, Counting, Grouped Summary statistics
 * Ch 3, Subsetting using slicing, Indexing and subsetting using index
* Data Visualization  
 * Ch 4, Plotting, Handling missing data, Reading data into a DataFrame
  
  
* **Introducing DataFrames**  
pandas is built on top of NumPy and Matplotlib.  
  
Rectangular data or tabular data is the a common form of data pandas is design to work with it.  
Each row is an observation and each column is a property or attribute. In pandas this rectangular data is represented as a `DataFrame` object. Other programming languages has something similar, `SQL` has `data table` and `R` also has `data.frame`. Every column have the same data type on each row but different columns can have different data types.  
  
Exploring a DataFrame object using methods/attributes:  
  
`df.head()` method displays the first rows (default is 5)  
`df.indo()` method displays the names of columns, data types they contain and if there are any missing alues.  
`df.shape` attribute shows the number of rows followed by the number of clolumns, e.g `(8,7)`  
`df.describe()` method calculates some summary statistics for numerical columns.  
  
A `DataFrame` consists of three different components accessible using attributes.  
  
`df.values` attribute contains the data values in a 2D numpy array  
`df.columns` attribute contains the labels of the column names as index object.  
`df.index` attribute contains row numbers of row names as index object. Note: Row labes are stored in `.index` and not in `.rows`!  
  
**pandas Philosophy**  
There should be one -- and preferably only one -- obvious way to do it!  
*- The Zen of Python* by Tim Peters, Item 13 https://www.python.org/dev/peps/pep-0020/ 
Pandas deliberately does not follow this philosphoy, instead there are often many ways to solve a problem in pandas and we can choose the best one. pandas gives many tools to give options.

**Inspecting a DataFrame**  
When we have a new DataFrame to work with, the first thing you need to do is explore it and see what it contains. There are several useful methods and attributes for this.  
  
`.head()` returns the first few rows (the “head” of the `DataFrame`).  
`.info()` shows information on each of the columns, such as the data type and number of missing values.  
`.shape` returns the number of rows and columns of the DataFrame.  
`.describe()` calculates a few summary statistics for each column.

In [11]:
import pandas as pd

# OMA Need to create the DF
path = "data/homelessness.csv"
homelessness = pd.read_csv(path, index_col=0)

In [12]:
# Print the head of the homelessness data
print(homelessness.head())

               region       state  individuals  family_members  state_pop
0  East South Central     Alabama       2570.0           864.0    4887681
1             Pacific      Alaska       1434.0           582.0     735139
2            Mountain     Arizona       7259.0          2606.0    7158024
3  West South Central    Arkansas       2280.0           432.0    3009733
4             Pacific  California     109008.0         20964.0   39461588


In [3]:
# Print information about homelessness
print(homelessness.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      51 non-null     int64  
 1   region          51 non-null     object 
 2   state           51 non-null     object 
 3   individuals     51 non-null     float64
 4   family_members  51 non-null     float64
 5   state_pop       51 non-null     int64  
dtypes: float64(2), int64(2), object(2)
memory usage: 2.5+ KB
None


In [4]:
# Print the shape of homelessness
print(homelessness.shape)

(51, 6)


In [6]:
# Print a description of homelessness
print(homelessness.describe())

       Unnamed: 0    individuals  family_members     state_pop
count   51.000000      51.000000       51.000000  5.100000e+01
mean    25.000000    7225.784314     3504.882353  6.405637e+06
std     14.866069   15991.025083     7805.411811  7.327258e+06
min      0.000000     434.000000       75.000000  5.776010e+05
25%     12.500000    1446.500000      592.000000  1.777414e+06
50%     25.000000    3082.000000     1482.000000  4.461153e+06
75%     37.500000    6781.500000     3196.000000  7.340946e+06
max     50.000000  109008.000000    52070.000000  3.946159e+07


We see the average number of homeless individuals of each state is about 7226.  
Now we do some more exploration of the df!  
  
**Parts of a DataFrame**  
To better understand DataFrame objects, it's useful to know that they consist of three components, stored as attributes:  
  
* `.values`: A two-dimensional NumPy array of values.  
* `.columns`: An index of columns: the column names.  
* `.index`: An index for the rows: either row numbers or row names.  
  
You can usually think of indexes as a list of strings or numbers, though the pandas `Index` data type allows for more sophisticated options. (These will be covered later in the course.)  
  
* Print a 2D NumPy array of the values in `homelessness`
* Print the column names of `homelessness`
* Print the index of `homelessness`

In [13]:
# Print the values of homelessness
print(homelessness.values)

[['East South Central' 'Alabama' 2570.0 864.0 4887681]
 ['Pacific' 'Alaska' 1434.0 582.0 735139]
 ['Mountain' 'Arizona' 7259.0 2606.0 7158024]
 ['West South Central' 'Arkansas' 2280.0 432.0 3009733]
 ['Pacific' 'California' 109008.0 20964.0 39461588]
 ['Mountain' 'Colorado' 7607.0 3250.0 5691287]
 ['New England' 'Connecticut' 2280.0 1696.0 3571520]
 ['South Atlantic' 'Delaware' 708.0 374.0 965479]
 ['South Atlantic' 'District of Columbia' 3770.0 3134.0 701547]
 ['South Atlantic' 'Florida' 21443.0 9587.0 21244317]
 ['South Atlantic' 'Georgia' 6943.0 2556.0 10511131]
 ['Pacific' 'Hawaii' 4131.0 2399.0 1420593]
 ['Mountain' 'Idaho' 1297.0 715.0 1750536]
 ['East North Central' 'Illinois' 6752.0 3891.0 12723071]
 ['East North Central' 'Indiana' 3776.0 1482.0 6695497]
 ['West North Central' 'Iowa' 1711.0 1038.0 3148618]
 ['West North Central' 'Kansas' 1443.0 773.0 2911359]
 ['East South Central' 'Kentucky' 2735.0 953.0 4461153]
 ['West South Central' 'Louisiana' 2540.0 519.0 4659690]
 ['New 

In [14]:
# Print the column index of homelessness
print(homelessness.columns)

Index(['region', 'state', 'individuals', 'family_members', 'state_pop'], dtype='object')


In [15]:
# Print the row index of homelessness
print(homelessness.index)

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49,
            50],
           dtype='int64')


A `DataFrame` has three components: values, a column index, and a row index  
  
**Sorting and subsetting**  
  
Sorting - change the order of the row by sorting them.  
  
`df.sort_values("col_name")` a method that takes column_name as argument  
`df.sort_values(["col_name_1", "col_name_2"], ascending = [True, False])` arguments for multiple columns and sorting order
`df["col_name"]`, to look at only one column
`df[["col_name_1"],["col_name_2"]]`, the outer squared brackets are responsible for subsetting the data frame, the inner squared brackets are responsible for creating a list of column names to subset.  
`cols_to_subset = ["col_name_1", "col_name_2"]` , create the list of columns as a variable   
`df[cols_to_subset]`, use the separate list as a variable to subset the df  
  
Subsetting rows - by creating logic to filter by  
`df[df["col_name_1"] > 50]` , subset the df with a logic of a column. We can sub-set also with text and dates  
  
Subsetting using multiple conditions using operators:  
`is_yes = df["col_name_1"] == "yes"`  
`is_no = df["col_name_2"] == "no"`  
`df[is_yes & is_no]`  
  
Or as one line of code, need paranthesis around each condition:  
`df[ (df["col_name_1"] == "yes") & (df["col_name_1"] == "no") ]`  
  
Subsetting using categorical values using `isin()` method:  
`is_black_or_brown = df["color"].isin(["Black", "Brown"])`  
`df[is_black_or_brown]`  

**Sorting rows**  
Examples

In [16]:
homelessness.head()

region       state  individuals  family_members  state_pop
0  East South Central     Alabama       2570.0           864.0    4887681
1             Pacific      Alaska       1434.0           582.0     735139
2            Mountain     Arizona       7259.0          2606.0    7158024
3  West South Central    Arkansas       2280.0           432.0    3009733
4             Pacific  California     109008.0         20964.0   39461588

Sort `homelessness` by the number of homeless `individuals` in ascending order, and save this as `homelessness_ind`.

In [17]:
# Sort homelessness by individual
homelessness_ind = homelessness.sort_values(["individuals"])

# Print the top few rows
print(homelessness_ind.head())

                region         state  individuals  family_members  state_pop
50            Mountain       Wyoming        434.0           205.0     577601
34  West North Central  North Dakota        467.0            75.0     758080
7       South Atlantic      Delaware        708.0           374.0     965479
39         New England  Rhode Island        747.0           354.0    1058287
45         New England       Vermont        780.0           511.0     624358


Sort `homelessness` by the number of homeless `family_members` in descending order, and save this as `homelessness_fam`.

In [18]:
# Sort homelessness by descending family members
homelessness_fam = homelessness.sort_values(["family_members"], ascending = False)

# Print the top few rows
print(homelessness_fam.head())

                region          state  individuals  family_members  state_pop
32        Mid-Atlantic       New York      39827.0         52070.0   19530351
4              Pacific     California     109008.0         20964.0   39461588
21         New England  Massachusetts       6811.0         13257.0    6882635
9       South Atlantic        Florida      21443.0          9587.0   21244317
43  West South Central          Texas      19199.0          6111.0   28628666


Sort `homelessness` first by region (ascending), and then by number of family members (descending). Save this as `homelessness_reg_fam`.

In [19]:
# Sort homelessness by region, then descending family members
homelessness_reg_fam = homelessness.sort_values(["region", "family_members"], ascending = [True, False])

# Print the top few rows
print(homelessness_reg_fam.head())

                region      state  individuals  family_members  state_pop
13  East North Central   Illinois       6752.0          3891.0   12723071
35  East North Central       Ohio       6929.0          3320.0   11676341
22  East North Central   Michigan       5209.0          3142.0    9984072
49  East North Central  Wisconsin       2740.0          2167.0    5807406
14  East North Central    Indiana       3776.0          1482.0    6695497


Using `.sort_values()` and `.head()` is a powerful pair.

**Subsetting columns**  
When working with data, you may not need all of the variables in your dataset. Square brackets (`[]`) can be used to select only the columns that matter to you in an order that makes sense to you. To select only `"col_a"` of the DataFrame `df`, use  
`df["col_a"]`  
To select two columns use  
`df[["col_a","col_b"]]`  

In [20]:
# Select the individuals column
individuals = homelessness["individuals"]

# Print the head of the result
print(individuals.head())

0      2570.0
1      1434.0
2      7259.0
3      2280.0
4    109008.0
Name: individuals, dtype: float64


In [21]:
# Select the state and family_members columns
state_fam = homelessness[["state","family_members"]]

# Print the head of the result
print(state_fam.head())

        state  family_members
0     Alabama           864.0
1      Alaska           582.0
2     Arizona          2606.0
3    Arkansas           432.0
4  California         20964.0


In [22]:
# Select only the individuals and state columns, in that order
ind_state = homelessness[["individuals","state"]]

# Print the head of the result
print(ind_state.head())

   individuals       state
0       2570.0     Alabama
1       1434.0      Alaska
2       7259.0     Arizona
3       2280.0    Arkansas
4     109008.0  California


**Subsetting rows**  
A large part of data science is about finding which bits of your dataset are interesting. One of the simplest techniques for this is to find a subset of rows that match some criteria. This is sometimes known as * filtering rows*  or *selecting rows*.  
  
There are many ways to subset a DataFrame, perhaps the most common is to use relational operators to return `True` or `False` for each row, then pass that inside square brackets.  
  
`df[df["height_cm"] > 60]`  
`df[df["color"] == "tan"]`  
  
We can filter for multiple conditions at once by using the "bitwise and" operator, `&`.  
  
`df[(df["height_cm"] > 60) & (df["color"] == "tan")]`

In [23]:
# Filter for rows where individuals is greater than 10000
ind_gt_10k = homelessness[homelessness["individuals"] > 10000]

# See the result
print(ind_gt_10k)

                region       state  individuals  family_members  state_pop
4              Pacific  California     109008.0         20964.0   39461588
9       South Atlantic     Florida      21443.0          9587.0   21244317
32        Mid-Atlantic    New York      39827.0         52070.0   19530351
37             Pacific      Oregon      11139.0          3337.0    4181886
43  West South Central       Texas      19199.0          6111.0   28628666
47             Pacific  Washington      16424.0          5880.0    7523869


In [24]:
# Filter for rows where region is Mountain
mountain_reg = homelessness[homelessness["region"] == "Mountain"]

# See the result
print(mountain_reg)

      region       state  individuals  family_members  state_pop
2   Mountain     Arizona       7259.0          2606.0    7158024
5   Mountain    Colorado       7607.0          3250.0    5691287
12  Mountain       Idaho       1297.0           715.0    1750536
26  Mountain     Montana        983.0           422.0    1060665
28  Mountain      Nevada       7058.0           486.0    3027341
31  Mountain  New Mexico       1949.0           602.0    2092741
44  Mountain        Utah       1904.0           972.0    3153550
50  Mountain     Wyoming        434.0           205.0     577601


In [25]:
# Filter for rows where family_members is less than 1000 
# and region is Pacific
fam_lt_1k_pac = homelessness[ (homelessness["family_members"] < 1000) & (homelessness["region"]=="Pacific") ]

# See the result
print(fam_lt_1k_pac)

    region   state  individuals  family_members  state_pop
1  Pacific  Alaska       1434.0           582.0     735139


Using square brackets plus logical conditions is often the most powerful way of identifying interesting rows of data.

**Subsetting rows by categorical variables**  
Subsetting data based on a categorical variable often involves using the "or" operator (`|`) to select rows from multiple categories. This can get tedious when you want all states in one of three different regions, for example. Instead, use the `.isin()` method, which will allow you to tackle this problem by writing one condition instead of three separate ones.  
  
`colors = ["brown", "black", "tan"]`  
`condition = df["color"].isin(colors)`  
`df[condition]`

In [26]:
homelessness.head()

region       state  individuals  family_members  state_pop
0  East South Central     Alabama       2570.0           864.0    4887681
1             Pacific      Alaska       1434.0           582.0     735139
2            Mountain     Arizona       7259.0          2606.0    7158024
3  West South Central    Arkansas       2280.0           432.0    3009733
4             Pacific  California     109008.0         20964.0   39461588

In [27]:
# Subset for rows in South Atlantic or Mid-Atlantic regions
south_mid_atlantic = homelessness[homelessness["region"].isin(["South Atlantic","Mid-Atlantic"])]

# See the result
print(south_mid_atlantic)

            region                 state  individuals  family_members  \
7   South Atlantic              Delaware        708.0           374.0   
8   South Atlantic  District of Columbia       3770.0          3134.0   
9   South Atlantic               Florida      21443.0          9587.0   
10  South Atlantic               Georgia       6943.0          2556.0   
20  South Atlantic              Maryland       4914.0          2230.0   
30    Mid-Atlantic            New Jersey       6048.0          3350.0   
32    Mid-Atlantic              New York      39827.0         52070.0   
33  South Atlantic        North Carolina       6451.0          2817.0   
38    Mid-Atlantic          Pennsylvania       8163.0          5349.0   
40  South Atlantic        South Carolina       3082.0           851.0   
46  South Atlantic              Virginia       3928.0          2047.0   
48  South Atlantic         West Virginia       1021.0           222.0   

    state_pop  
7      965479  
8      701547  
9 

In [28]:
# The Mojave Desert states
canu = ["California", "Arizona", "Nevada", "Utah"]

# Filter for rows in the Mojave Desert states
mojave_homelessness = homelessness[homelessness["state"].isin(canu)]

# See the result
print(mojave_homelessness)

      region       state  individuals  family_members  state_pop
2   Mountain     Arizona       7259.0          2606.0    7158024
4    Pacific  California     109008.0         20964.0   39461588
28  Mountain      Nevada       7058.0           486.0    3027341
44  Mountain        Utah       1904.0           972.0    3153550


Using `.isin()` makes subsetting categorical variables a breeze  
  
**New Columns**  
  
Often we want to add new columns derived from existing columns. Adding a new column can go by many names like mutating a data frame, transforming a data frame or feature engineering.  
  
Example of how to add a new column from existing columns:  
`df["heigh_m"] = df["height_cm"]  / 100`  
On the left hand side we have in square brackets the name of the column we want to create.  
On the right hand side we have the calculation.  
We will then have both the existing and the new column in the data frame.  
  
**Multiple manipulations**  
Skinny dogs e.g  
`bmi_lt_100 = df[df["bmi] < 100]`  
`bmi_lt_100_height = bmi_lt_100.sort_values("height_cm", ascending=False)`  
`bmi_lt_100_height[["name", "height_cm", "bmi"]]`  
  
Some examples:

**Adding new columns**  
You aren't stuck with just the data you are given. Instead, you can add new columns to a DataFrame. This has many names, such as *transforming*, *mutating*, and *feature engineering*.  
  
You can create new columns from scratch, but it is also common to derive them from other columns, for example, by adding columns together or by changing their units.

In [29]:
# Add total col as sum of individuals and family_members
homelessness["total"] = homelessness["individuals"] + homelessness["family_members"]

# Add p_individuals col as proportion of individuals
homelessness["p_individuals"] = homelessness["individuals"] / homelessness["total"]

# See the result
print(homelessness.head())

               region       state  individuals  family_members  state_pop  \
0  East South Central     Alabama       2570.0           864.0    4887681   
1             Pacific      Alaska       1434.0           582.0     735139   
2            Mountain     Arizona       7259.0          2606.0    7158024   
3  West South Central    Arkansas       2280.0           432.0    3009733   
4             Pacific  California     109008.0         20964.0   39461588   

      total  p_individuals  
0    3434.0       0.748398  
1    2016.0       0.711310  
2    9865.0       0.735834  
3    2712.0       0.840708  
4  129972.0       0.838704  


If our dataset doesn't have the exact columns you need, we can often make your own from what you have.  
  
**Combo-attack!**  
You've seen the four most common types of data manipulation: sorting rows, subsetting columns, subsetting rows, and adding new columns. In a real-life data analysis, you can mix and match these four manipulations to answer a multitude of questions.  
  
In this exercise, you'll answer the question, "Which state has the highest number of homeless individuals per 10,000 people in the state?"  

In [30]:
# Create indiv_per_10k col as homeless individuals per 10k state pop
homelessness["indiv_per_10k"] = 10000 * homelessness["individuals"] / homelessness["state_pop"] 

# Subset rows for indiv_per_10k greater than 20
high_homelessness = homelessness[homelessness["indiv_per_10k"] > 20]

# Sort high_homelessness by descending indiv_per_10k
high_homelessness_srt = high_homelessness.sort_values("indiv_per_10k", ascending=False) 

# From high_homelessness_srt, select the state and indiv_per_10k cols
result = high_homelessness_srt[["state","indiv_per_10k"]]

# See the result
print(result)

                   state  indiv_per_10k
8   District of Columbia      53.738381
11                Hawaii      29.079406
4             California      27.623825
37                Oregon      26.636307
28                Nevada      23.314189
47            Washington      21.829195
32              New York      20.392363


Washington, D.C. has the highest number of homeless individuals - almost 54 per ten thousand people. This is almost double the number of the next-highest state, Hawaii. If you combine new column addition, row subsetting, sorting, and column selection, you can answer lots of questions like this.  
  
### 2. Aggregating DataFrames  
  
**Summary statistics**  
Focus here is on aggregating data using summary statistics. We will use dot-notation. E.g.  
`df["col_1"].mean()`  
There are many more summary statistics that we can calculate like:  
`.median()`  
`.mode()`  
`.min()`  
`.max()`  
`.vat()`  
`.std()`  
We can also take sums and calculate quantiles:  
`.sum()`  
`.quantile()`  
  
The .agg() methods  
  
Allows for calculating custom summary statistics.  
`def pct30(column)`  
` return column.quantile(0.3)`  
Calculate the 30th percentile of df columns  
`df["col_1"].agg(pct30)` 
  
Summaries on multiple columns  
`df[["col_1", "col_2"]].agg(pct30)`  
  
Multiple summaries  
`def pct40(column)`  
` return column.quantile(0.4)`  
Calculate the 30th and 40th percentile of df column  
`df["col_1"].agg([pct30, pct40])`  
  
Cumulative sum  
`df["col_1].cumsum()`
It returns one sum for each row of a df.  
Example see below:

In [1]:
import pandas as pd

# OMA Need to create the DF
path = "data/homelessness.csv"
homelessness = pd.read_csv(path, index_col=0)

homelessness.head()

region       state  individuals  family_members  state_pop
0  East South Central     Alabama       2570.0           864.0    4887681
1             Pacific      Alaska       1434.0           582.0     735139
2            Mountain     Arizona       7259.0          2606.0    7158024
3  West South Central    Arkansas       2280.0           432.0    3009733
4             Pacific  California     109008.0         20964.0   39461588

In [4]:
# cumsum
homelessness["ind_cumsum"] = homelessness["individuals"].cumsum()
homelessness[["individuals","ind_cumsum"]].head()

individuals  ind_cumsum
0       2570.0      2570.0
1       1434.0      4004.0
2       7259.0     11263.0
3       2280.0     13543.0
4     109008.0    122551.0

Other Cumulative Statistics  
`.cummax()`  
`.cummin()`  
`.cumprod()`  
All these return an entire column rather than one value.  
  
**Mean and median**  
Some examples

In [5]:
# OMA need to read in the data first
path2 = "data/sales_subset.csv"
sales = pd.read_csv(path2, index_col=0)
sales.head()

store type  department        date  weekly_sales  is_holiday  \
0      1    A           1  2010-02-05      24924.50       False   
1      1    A           1  2010-03-05      21827.90       False   
2      1    A           1  2010-04-02      57258.43       False   
3      1    A           1  2010-05-07      17413.94       False   
4      1    A           1  2010-06-04      17558.09       False   

   temperature_c  fuel_price_usd_per_l  unemployment  
0       5.727778              0.679451         8.106  
1       8.055556              0.693452         8.106  
2      16.816667              0.718284         7.808  
3      22.527778              0.748928         7.808  
4      27.050000              0.714586         7.808

In [6]:
# Print the head of the sales DataFrame
print(sales.head())

# Print the info about the sales DataFrame
print(sales.info())

# Print the mean of weekly_sales
print(sales["weekly_sales"].mean())

# Print the median of weekly_sales
print(sales["weekly_sales"].median())

   store type  department        date  weekly_sales  is_holiday  \
0      1    A           1  2010-02-05      24924.50       False   
1      1    A           1  2010-03-05      21827.90       False   
2      1    A           1  2010-04-02      57258.43       False   
3      1    A           1  2010-05-07      17413.94       False   
4      1    A           1  2010-06-04      17558.09       False   

   temperature_c  fuel_price_usd_per_l  unemployment  
0       5.727778              0.679451         8.106  
1       8.055556              0.693452         8.106  
2      16.816667              0.718284         7.808  
3      22.527778              0.748928         7.808  
4      27.050000              0.714586         7.808  
<class 'pandas.core.frame.DataFrame'>
Int64Index: 10774 entries, 0 to 10773
Data columns (total 9 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   store                 10774 non-null  int64  
 1   t

The mean weekly sales amount is almost double the median weekly sales amount! This can tell you that there are a few very high sales weeks that are making the mean so much higher than the median.  
  
**Summarizing dates**  
Summary statistics can also be calculated on date columns that have values with the data type `datetime64`. Some summary statistics — like mean — don't make a ton of sense on dates, but others are super helpful, for example, minimum and maximum, which allow you to see what time range your data covers.

In [7]:
# Print the maximum of the date column
print(sales["date"].max())

# Print the minimum of the date column
print(sales["date"].min())

2012-10-26
2010-02-05


Taking the minimum and maximum of a column of dates is handy for figuring out what time period your data covers. In this case, there are data from February of 2010 to October of 2012.  
  
**Efficient summaries**  
While pandas and NumPy have tons of functions, sometimes, you may need a different function to summarize your data.  
  
The `.agg()` method allows you to apply your own custom functions to a DataFrame, as well as apply functions to more than one column of a DataFrame at once, making your aggregations super-efficient. For example,  
`df['column'].agg(function)`  

In [8]:
# A custom IQR function
def iqr(column):
    return column.quantile(0.75) - column.quantile(0.25)
    
# Print IQR of the temperature_c column
print(sales["temperature_c"].agg(iqr))

16.58333333333334


In [9]:
# A custom IQR function
def iqr(column):
    return column.quantile(0.75) - column.quantile(0.25)

# Update to print IQR of temperature_c, fuel_price_usd_per_l, & unemployment
print(sales[["temperature_c", "fuel_price_usd_per_l", "unemployment"]].agg(iqr))

temperature_c           16.583333
fuel_price_usd_per_l     0.073176
unemployment             0.565000
dtype: float64


In [10]:
# Import NumPy and create custom IQR function
import numpy as np
def iqr(column):
    return column.quantile(0.75) - column.quantile(0.25)

# Update to print IQR and median of temperature_c, fuel_price_usd_per_l, & unemployment
print(sales[["temperature_c", "fuel_price_usd_per_l", "unemployment"]].agg([iqr,np.median]))

        temperature_c  fuel_price_usd_per_l  unemployment
iqr         16.583333              0.073176         0.565
median      16.966667              0.743381         8.099


The `.agg()` method makes it easy to compute multiple statistics on multiple columns, all in just one line of code.  
  
**Cumulative statistics**  
Cumulative statistics can also be helpful in tracking summary statistics over time. In this exercise, you'll calculate the cumulative sum and cumulative max of a department's weekly sales, which will allow you to identify what the total sales were so far as well as what the highest weekly sales were so far.

In [24]:
# Create dataset needed. Only include store = 1 and department = 1
sales_1_1 = sales[ (sales["store"]==1) & (sales["department"]==1) ]
sales_1_1.shape

(144, 9)

In [25]:
sales_1_1.head()

store type  department        date  weekly_sales  is_holiday  \
0      1    A           1  2010-02-05      24924.50       False   
1      1    A           1  2010-03-05      21827.90       False   
2      1    A           1  2010-04-02      57258.43       False   
3      1    A           1  2010-05-07      17413.94       False   
4      1    A           1  2010-06-04      17558.09       False   

   temperature_c  fuel_price_usd_per_l  unemployment  
0       5.727778              0.679451         8.106  
1       8.055556              0.693452         8.106  
2      16.816667              0.718284         7.808  
3      22.527778              0.748928         7.808  
4      27.050000              0.714586         7.808

In [27]:
# Sort sales_1_1 by date
sales_1_1 = sales_1_1.sort_values("date").head(12)

# Get the cumulative sum of weekly_sales, add as cum_weekly_sales col
sales_1_1["cum_weekly_sales"] = sales_1_1["weekly_sales"].cumsum()

# Get the cumulative max of weekly_sales, add as cum_max_sales col
sales_1_1["cum_max_sales"] = sales_1_1["weekly_sales"].cummax()

# See the columns you calculated
print(sales_1_1[["date", "weekly_sales", "cum_weekly_sales", "cum_max_sales"]])

            date  weekly_sales  cum_weekly_sales  cum_max_sales
0     2010-02-05      24924.50          24924.50       24924.50
9009  2010-02-05      18187.71          43112.21       24924.50
8109  2010-02-05      32313.79          75426.00       32313.79
7199  2010-02-05      46021.21         121447.21       46021.21
6293  2010-02-05      21500.58         142947.79       46021.21
5408  2010-02-05      32842.31         175790.10       46021.21
4495  2010-02-05      46761.90         222552.00       46761.90
3593  2010-02-05      40212.84         262764.84       46761.90
2699  2010-02-05      25619.00         288383.84       46761.90
1798  2010-02-05      38724.42         327108.26       46761.90
901   2010-02-05      35034.06         362142.32       46761.90
9899  2010-02-05      21244.50         383386.82       46761.90


Not all functions that calculate on columns return a single number. Some, like the cumulative statistic functions, return a whole column.  
  
**Counting**  
  
Now we will summarize categorical data using counting.  
  
Doprring duplicate names  
`df.drop_duplicates(subset="col_name")`  
  
This method takes an argument, the column_name, we want to find our duplicates based on. In this case we want all the unique names. We can drop based on two columns.  
`df.drop_duplicates(subset=["col_name_1","col_name_2])`  
  
Now we base our duplicate dropping on multiple columns by passing a list of column names into the `subset` argument. Now for example if we have string `Tom` twice in the `col_name_1` but with `Wattson` and `Johnson` in `col_name_2` we can thus keep duplicates in the `col_name_1` since it is the combination of both columns that needs to be unique.  
  
To count the rows in each `col_name_1` we can  
`df["col_name_1"].value_counts(sort=True)` subset the df by col_name_1 and count the rows and sort the output. We can also use the `normalize` argument to turn the counts into proportions of total.  
`df["col_name_1"].value_counts(normalize=True)`  
  
Some exaxmples.  
  
**Dropping duplicates**  
Removing duplicates is an essential skill to get accurate counts because often, you don't want to count the same thing multiple times. In this exercise, you'll create some new DataFrames using unique values from `sales`.

In [30]:
# Drop duplicate store/type combinations
store_types = sales.drop_duplicates(subset=["store", "type"])
print(store_types.head())

# Drop duplicate store/department combinations
store_depts = sales.drop_duplicates(subset=["store", "department"])
print(store_depts.head())

# Subset the rows where is_holiday is True and drop duplicate dates
holiday_dates = sales[sales["is_holiday"]].drop_duplicates(subset="date")

# Print date col of holiday_dates
print(holiday_dates["date"])

      store type  department        date  weekly_sales  is_holiday  \
0         1    A           1  2010-02-05      24924.50       False   
3593      1    B           1  2010-02-05      40212.84       False   

      temperature_c  fuel_price_usd_per_l  unemployment  
0          5.727778              0.679451         8.106  
3593      12.411111              0.782478         9.765  
    store type  department        date  weekly_sales  is_holiday  \
0       1    A           1  2010-02-05      24924.50       False   
12      1    A           2  2010-02-05      50605.27       False   
24      1    A           3  2010-02-05      13740.12       False   
36      1    A           4  2010-02-05      39954.04       False   
48      1    A           5  2010-02-05      32229.38       False   

    temperature_c  fuel_price_usd_per_l  unemployment  
0        5.727778              0.679451         8.106  
12       5.727778              0.679451         8.106  
24       5.727778              0.67945

The holiday weeks correspond to the Superbowl in February, Labor Day in September, Thanksgiving in November, and Christmas in December. Now that the duplicates are removed, it's time to do some counting.  
  
**Counting categorical variables**  
Counting is a great way to get an overview of your data and to spot curiosities that you might not notice otherwise. In this exercise, you'll count the number of each type of store and the number of each department number using the DataFrames you created in the previous exercise:  
  
`# Drop duplicate store/type combinations`  
`store_types = sales.drop_duplicates(subset=["store", "type"])`  
  
`# Drop duplicate store/department combinations`  
`store_depts = sales.drop_duplicates(subset=["store", "department"])`

In [32]:
# Count the number of stores of each type
store_counts = store_types["type"].value_counts()
print(store_counts)

# Get the proportion of stores of each type
store_props = store_types["type"].value_counts(normalize=True)
print(store_props)

# Count the number of each department number and sort
dept_counts_sorted = store_depts["department"].value_counts()
print(dept_counts_sorted)

# Get the proportion of departments of each number and sort
dept_props_sorted = store_depts["department"].value_counts(sort=True, normalize=True)
print(dept_props_sorted)

B    1
A    1
Name: type, dtype: int64
B    0.5
A    0.5
Name: type, dtype: float64
99    1
98    1
23    1
24    1
25    1
     ..
58    1
59    1
60    1
67    1
1     1
Name: department, Length: 80, dtype: int64
99    0.0125
98    0.0125
23    0.0125
24    0.0125
25    0.0125
       ...  
58    0.0125
59    0.0125
60    0.0125
67    0.0125
1     0.0125
Name: department, Length: 80, dtype: float64


**Grouped summary statistics**  
TBD

In [ ]:
#https://campus.datacamp.com/courses/data-manipulation-with-pandas/aggregating-dataframes?ex=9